File: components/search_engines/android/java/src/org/chromium/components/search_engines/SearchEnginesFeatureMap.java
// Copyright 2024 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.components.search_engines;

import org.jni_zero.JNINamespace;
import org.jni_zero.NativeMethods;

import org.chromium.base.FeatureMap;

/**
 * Java accessor for search engines base::Features.
 *
 * <p>Note: Features must be added to the array `kFeaturesExposedToJava` in
 * //components/search_engines/android/search_engines_feature_map.cc
 */
@JNINamespace("search_engines")
public final class SearchEnginesFeatureMap extends FeatureMap {
    private static final SearchEnginesFeatureMap sInstance = new SearchEnginesFeatureMap();

    // Do not instantiate this class.
    private SearchEnginesFeatureMap() {}

    /**
     * @return the singleton SearchEnginesFeatureMap.
     */
    public static SearchEnginesFeatureMap getInstance() {
        return sInstance;
    }

    /** Convenience method to call {@link #isEnabledInNative(String)} statically. */
    public static boolean isEnabled(String featureName) {
        return getInstance().isEnabledInNative(featureName);
    }

    @Override
    protected long getNativeMap() {
        return SearchEnginesFeatureMapJni.get().getNativeMap();
    }

    @NativeMethods
    public interface Natives {
        long getNativeMap();
    }
}

File: components/search_engines/android/search_engines_feature_map.cc
// Copyright 2024 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <iterator>
#include <string>

#include "base/android/feature_map.h"
#include "base/feature_list.h"
#include "base/features.h"
#include "base/no_destructor.h"
#include "components/search_engines/search_engines_switches.h"

// Must come after all headers that specialize FromJniType() / ToJniType().
#include "components/search_engines/android/features_jni_headers/SearchEnginesFeatureMap_jni.h"

namespace search_engines {

namespace {

// Array of features exposed through the Java BaseFeatureMap API. Entries in
// this array refer to features defined in //search_engines features.
const base::Feature* const kFeaturesExposedToJava[] = {
    &switches::kClayBlocking};

// static
base::android::FeatureMap* GetFeatureMap() {
  static base::NoDestructor<base::android::FeatureMap> kFeatureMap(std::vector(
      std::begin(kFeaturesExposedToJava), std::end(kFeaturesExposedToJava)));
  return kFeatureMap.get();
}

}  // namespace

static jlong JNI_SearchEnginesFeatureMap_GetNativeMap(JNIEnv* env) {
  return reinterpret_cast<jlong>(GetFeatureMap());
}

}  // namespace search_engines
